flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaCommitter.java

```diff
@@ -64,19 +65,31 @@
                                 .<FlinkKafkaInternalProducer<?, ?>>map(Recyclable::getObject)
                                 .orElseGet(() -> getRecoveryProducer(committable));
                 producer.commitTransaction();
-                recyclable.ifPresent(Recyclable::close);
-            } catch (ProducerFencedException | InvalidTxnStateException e) {
-                // That means we have committed this transaction before.
+            } catch (RetriableException e) {
                 LOG.warn(
-                        "Encountered error {} while recovering transaction {}. "
-                                + "Presumably this transaction has been already committed before",
-                        e,
-                        committable);
-                recyclable.ifPresent(Recyclable::close);
-            } catch (Throwable e) {
-                LOG.warn("Cannot commit Kafka transaction, retrying.", e);
+                        "Encountered retriable exception while committing {}.", transactionalId, e);
                 retryableCommittables.add(committable);
+                continue;
+            } catch (ProducerFencedException e) {
+                // initTransaction has been called on this transaction before
+                LOG.error(
+                        "Transactions {} timed out or was overridden and data has been potentially lost.",
+                        transactionalId,
+                        e);
+            } catch (InvalidTxnStateException e) {
+                // This exception only occurs when aborting after a commit or vice versa.
+                // It does not appear on double commits or double aborts.
+                LOG.error(
+                        "Transaction {} was previously canceled and data has been lost.",
+                        committable,
+                        e);
+            } catch (Exception e) {
+                LOG.error(
+                        "Transaction {} encountered error and data has been potentially lost.",
+                        committable,
+                        e);
             }
+            recyclable.ifPresent(Recyclable::close);
```

Comment:
> Why can we close the producer if we might want to retry committing the Committables? 🤔
We are not retrying them at this point. Retriable path is short-cutted with `continue`.

Reply:
True. IMHO, the code is hard to understand here because of the little `continue` being hidden in the catch blocks. But I cannot come up with a quick fix. ¯\_(ツ)_/¯